File: tests/test_browser.js
/**
 * NEO Voice — Browser UI Test via Playwright (Node.js)
 * Automated: navigate → select voice → enter text → create MP3 → verify
 */

const fs = require("fs");
const path = require("path");

const SCREENSHOTS = path.join(__dirname, "screenshots");
fs.mkdirSync(SCREENSHOTS, { recursive: true });

async function run() {
  let chromium;
  try {
    chromium = require("playwright").chromium;
  } catch {
    try {
      chromium = require("playwright-core").chromium;
    } catch {
      console.error("❌ Playwright not found. Install: npm i -g playwright");
      process.exit(1);
    }
  }

  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage({ viewport: { width: 1400, height: 900 } });

  try {
    // Step 1: Navigate
    console.log("1️⃣  Navigating to http://localhost:5173...");
    await page.goto("http://localhost:5173", { waitUntil: "networkidle", timeout: 15000 });
    await page.screenshot({ path: `${SCREENSHOTS}/01_main_page.png` });
    console.log(`   📸 01_main_page.png — Title: ${await page.title()}`);

    // Step 2: Navigate to Create Voice if needed
    const navLink = await page.$('a:has-text("Tạo"), a:has-text("giọng"), [href*="create"]');
    if (navLink) {
      console.log("   Clicking nav link to Create Voice...");
      await navLink.click();
      await page.waitForTimeout(1000);
    }
    await page.screenshot({ path: `${SCREENSHOTS}/02_create_page.png` });
    console.log("   📸 02_create_page.png");

    // Step 3: Select a VieNeu voice
    console.log("\n2️⃣  Looking for VieNeu voice...");
    const voiceNames = ["Bình An", "Ngọc Lan", "Xuân Vĩnh", "Gia Bảo", "Ngọc Linh"];
    let clicked = false;
    for (const name of voiceNames) {
      const el = await page.$(`text=${name}`);
      if (el) {
        await el.click();
        console.log(`   ✅ Selected voice: ${name}`);
        clicked = true;
        await page.waitForTimeout(500);
        break;
      }
    }
    if (!clicked) console.log("   ⚠️  No VieNeu voice found by name");
    await page.screenshot({ path: `${SCREENSHOTS}/03_voice_selected.png` });
    console.log("   📸 03_voice_selected.png");

    // Step 4: Enter text
    console.log("\n3️⃣  Entering text...");
    const textarea = await page.$("textarea");
    if (textarea) {
      await textarea.fill("Xin chào, đây là bài kiểm tra tự động. Hệ thống NEO Voice hoạt động tốt.");
      console.log("   ✅ Text entered");
    } else {
      console.log("   ❌ No textarea found!");
    }
    await page.screenshot({ path: `${SCREENSHOTS}/04_text_entered.png` });
    console.log("   📸 04_text_entered.png");

    // Step 5: Click Tạo MP3
    console.log("\n4️⃣  Clicking 'Tạo MP3'...");
    const renderBtn = await page.$('button:has-text("Tạo MP3")');
    if (renderBtn) {
      const disabled = await renderBtn.getAttribute("disabled");
      if (disabled === null) {
        await renderBtn.click();
        console.log("   ✅ Clicked 'Tạo MP3'");
      } else {
        console.log("   ⚠️  Button is disabled");
      }
    } else {
      console.log("   ❌ Button not found");
    }
    await page.screenshot({ path: `${SCREENSHOTS}/05_after_click.png` });
    console.log("   📸 05_after_click.png");

    // Step 6: Wait for completion
    console.log("\n5️⃣  Waiting for job (max 40s)...");
    let completed = false;
    for (let i = 0; i < 27; i++) {
      await page.waitForTimeout(1500);
      
      const success = await page.$('text=Hoàn thành');
      const audio = await page.$('audio[controls]');
      const failed = await page.$('text=thất bại');

      if (success || audio) {
        console.log(`   ✅ Job completed after ~${Math.round((i+1)*1.5)}s!`);
        completed = true;
        break;
      }
      if (failed) {
        const errEl = await page.$('[style*="error"], [style*="dc2626"]');
        const msg = errEl ? await errEl.innerText() : "unknown";
        console.log(`   ❌ Job failed: ${msg}`);
        break;
      }

      // Progress check
      if (i % 3 === 0) {
        const progEl = await page.$('.progress-info');
        if (progEl) {
          const text = await progEl.innerText();
          if (text.trim()) console.log(`   ⏳ ${text.trim()}`);
        }
      }
    }
    await page.screenshot({ path: `${SCREENSHOTS}/06_result.png` });
    console.log("   📸 06_result.png");

    // Step 7: Verify UI elements
    console.log("\n6️⃣  Verifying result UI...");
    const audioPlayer = await page.$("audio[controls]");
    const downloadBtn = await page.$("a[download]");
    const filenameEl = await page.$("[style*='monospace']");

    if (audioPlayer) {
      const src = await audioPlayer.getAttribute("src");
      console.log(`   ✅ Audio player: ${(src || "").substring(0, 80)}`);
    } else {
      console.log("   ❌ No audio player");
    }

    if (downloadBtn) {
      const href = await downloadBtn.getAttribute("href");
      console.log(`   ✅ Download button: ${(href || "").substring(0, 80)}`);
    } else {
      console.log("   ❌ No download button");
    }

    if (filenameEl) {
      console.log(`   ✅ Filename: ${await filenameEl.innerText()}`);
    } else {
      console.log("   ❌ No filename display");
    }

    // Full page screenshot
    await page.screenshot({ path: `${SCREENSHOTS}/07_final.png`, fullPage: true });
    console.log("   📸 07_final.png (full page)");

    console.log(completed ? "\n🎉 BROWSER TEST PASSED!" : "\n⚠️  BROWSER TEST INCOMPLETE");

  } catch (err) {
    console.error(`\n❌ Test error: ${err.message}`);
    await page.screenshot({ path: `${SCREENSHOTS}/error.png` }).catch(() => {});
  } finally {
    await browser.close();
  }
}

run();

File: tests/test_browser.py
"""
NEO Voice — Browser UI Test via Playwright
Automated test: navigate, select voice, enter text, create MP3, verify result
"""
import sys
import time

sys.stdout.reconfigure(encoding="utf-8", errors="replace")

from playwright.sync_api import sync_playwright

SCREENSHOTS_DIR = "D:/TQD-Voice/tests/screenshots"

import os
os.makedirs(SCREENSHOTS_DIR, exist_ok=True)

def run():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page(viewport={"width": 1400, "height": 900})

        # Step 1: Navigate
        print("1️⃣ Navigating to http://localhost:5173...")
        page.goto("http://localhost:5173", wait_until="networkidle", timeout=15000)
        page.screenshot(path=f"{SCREENSHOTS_DIR}/01_main_page.png", full_page=False)
        print(f"   📸 Screenshot: 01_main_page.png")
        print(f"   Title: {page.title()}")

        # Step 2: Check if we need to navigate to Create Voice page
        nav_links = page.query_selector_all("a, button, [role='tab']")
        for link in nav_links:
            text = (link.inner_text() or "").strip()
            if "tạo" in text.lower() or "create" in text.lower() or "giọng" in text.lower():
                print(f"   Found nav link: '{text}' — clicking...")
                link.click()
                page.wait_for_timeout(1000)
                break

        page.screenshot(path=f"{SCREENSHOTS_DIR}/02_create_voice_page.png", full_page=False)
        print(f"   📸 Screenshot: 02_create_voice_page.png")

        # Step 3: Find and click a VieNeu voice card
        print("\n2️⃣ Looking for VieNeu voice cards...")
        voice_cards = page.query_selector_all("[class*='voice'], [class*='card']")
        clicked_voice = False
        
        # Try clicking by text content containing VieNeu voice names
        vieneu_names = ["Bình An", "Ngọc Lan", "Xuân Vĩnh", "Gia Bảo", "Thái Sơn", "Ngọc Linh"]
        for name in vieneu_names:
            el = page.query_selector(f"text={name}")
            if el:
                print(f"   Found voice: '{name}' — clicking...")
                el.click()
                page.wait_for_timeout(500)
                clicked_voice = True
                break
        
        if not clicked_voice:
            # Fallback: try any clickable element with 'vieneu' text
            el = page.query_selector("text=vieneu")
            if el:
                el.click()
                page.wait_for_timeout(500)
                clicked_voice = True
                print("   Clicked element with 'vieneu' text")

        page.screenshot(path=f"{SCREENSHOTS_DIR}/03_voice_selected.png", full_page=False)
        print(f"   📸 Screenshot: 03_voice_selected.png")
        print(f"   Voice selected: {clicked_voice}")

        # Step 4: Enter text
        print("\n3️⃣ Entering text...")
        textarea = page.query_selector("textarea")
        if textarea:
            textarea.fill("Xin chào, đây là bài kiểm tra tự động từ trình duyệt. Hệ thống NEO Voice đang hoạt động tốt.")
            print("   ✅ Text entered in textarea")
        else:
            print("   ❌ No textarea found!")
        
        page.screenshot(path=f"{SCREENSHOTS_DIR}/04_text_entered.png", full_page=False)
        print(f"   📸 Screenshot: 04_text_entered.png")

        # Step 5: Click "Tạo MP3" button
        print("\n4️⃣ Clicking 'Tạo MP3' button...")
        render_btn = page.query_selector("button.btn-render, button:has-text('Tạo MP3')")
        if render_btn:
            is_disabled = render_btn.get_attribute("disabled")
            print(f"   Button found, disabled={is_disabled}")
            if not is_disabled:
                render_btn.click()
                print("   ✅ Clicked 'Tạo MP3'")
            else:
                print("   ⚠️ Button is disabled — checking why...")
                page.screenshot(path=f"{SCREENSHOTS_DIR}/04b_button_disabled.png")
        else:
            print("   ❌ 'Tạo MP3' button not found!")
            # Debug: list all buttons
            buttons = page.query_selector_all("button")
            for btn in buttons[:10]:
                print(f"      - Button: '{(btn.inner_text() or '').strip()[:50]}'")

        page.screenshot(path=f"{SCREENSHOTS_DIR}/05_after_click.png", full_page=False)
        print(f"   📸 Screenshot: 05_after_click.png")

        # Step 6: Wait for job to complete
        print("\n5️⃣ Waiting for job to complete (max 30s)...")
        for i in range(20):
            time.sleep(1.5)
            # Check for completion indicators
            success_el = page.query_selector("text=Hoàn thành")
            error_el = page.query_selector("text=thất bại")
            audio_el = page.query_selector("audio")
            
            if success_el or audio_el:
                print(f"   ✅ Job completed after ~{int((i+1)*1.5)}s!")
                break
            if error_el:
                error_text = page.query_selector("[style*='error'], [style*='red']")
                msg = error_text.inner_text() if error_text else "unknown error"
                print(f"   ❌ Job failed after ~{int((i+1)*1.5)}s: {msg}")
                break
            
            # Check progress
            progress_el = page.query_selector(".progress-info")
            if progress_el:
                prog_text = progress_el.inner_text().strip()
                if prog_text and i % 3 == 0:
                    print(f"   ⏳ Progress: {prog_text}")

        page.screenshot(path=f"{SCREENSHOTS_DIR}/06_result.png", full_page=False)
        print(f"   📸 Screenshot: 06_result.png")

        # Step 7: Check final state
        print("\n6️⃣ Final state check...")
        audio_player = page.query_selector("audio[controls]")
        download_btn = page.query_selector("a[download]")
        filename_el = page.query_selector("[style*='monospace']")
        
        if audio_player:
            src = audio_player.get_attribute("src") or ""
            print(f"   ✅ Audio player found, src={src[:80]}...")
        else:
            print("   ❌ No audio player found")

        if download_btn:
            href = download_btn.get_attribute("href") or ""
            print(f"   ✅ Download button found, href={href[:80]}...")
        else:
            print("   ❌ No download button found")

        if filename_el:
            fname = filename_el.inner_text().strip()
            print(f"   ✅ Filename shown: {fname}")
        else:
            print("   ❌ No filename display found")

        # Final full-page screenshot
        page.screenshot(path=f"{SCREENSHOTS_DIR}/07_final.png", full_page=True)
        print(f"   📸 Screenshot: 07_final.png (full page)")

        browser.close()
        print("\n✅ Browser test complete!")

if __name__ == "__main__":
    run()
